File: resources/engine/naming_hints.py
# -*- coding: utf-8 -*-
"""
Подсказки для переименования обфусцированных классов - НЕ угадывание, а
чтение уже существующих в байткоде данных (строк/аннотаций), которые
осмысленно называют класс. Идея и часть логики - с Rust-инструмента
пользователя (names.rs: moduleNames/commandNames), переписано на Python
под наши структуры данных (classfile.py).

1. by_annotation_name(): если МНОГО разных классов в jar'е несут одну и ту
   же (возможно, тоже обфусцированную по имени) аннотацию с текстовым
   аргументом "name" - это, скорее всего, самодельный DI/модульный
   фреймворк ("@Module(name=\"lobby\")" и т.п.) - берём это объявленное имя
   вместо генерического ClassA7.

2. by_brigadier_super_call(): специфично для Minecraft-плагинов на
   Brigadier (командный API Mojang, 1.13+) - командные классы обычно
   вызывают `super("имякоманды")` в конструкторе. Если проект использует
   Brigadier (есть упоминание в constant pool) и класс с обфусцированным
   именем расширяет ТОЖЕ обфусцированный класс, а его <init> начинается
   ровно с `aload_0; ldc "строка"; invokespecial <super>.<init>` - берём
   эту строку как имя команды.

Оба метода дают {internal_class_name: предложенное_имя} - используется
Renamer'ом (main.py) как ПОДСКАЗКА при переименовании, а не заменяет
обычную эвристику - если подсказки нет, поведение не меняется.
"""
import re

_VALID_IDENT_CHARS = re.compile(r"[^A-Za-z0-9_]")


def _sanitize_identifier(raw, prefix=""):
    """Превращает произвольную строку в валидный Java-идентификатор в
    PascalCase - 'teleport-player' -> 'TeleportPlayer', 'lobby' -> 'Lobby'."""
    parts = re.split(r"[^A-Za-z0-9]+", raw)
    parts = [p for p in parts if p]
    if not parts:
        return None
    name = "".join(p[:1].upper() + p[1:] for p in parts)
    name = _VALID_IDENT_CHARS.sub("", name)
    if not name or name[0].isdigit():
        name = "_" + name
    return prefix + name


def by_annotation_name(class_files, looks_obfuscated_fn):
    """cf.annotations - список {"type": descriptor, "args": {name: value}}
    (см. classfile.py). Считаем, сколько РАЗНЫХ классов несут каждый тип
    аннотации с аргументом "name" - если >= 3 разных классов (не случайное
    совпадение на одном классе), считаем это надёжным сигналом фреймворка."""
    by_annotation_type = {}  # descriptor -> list[(internal, name_value)]
    for internal, cf in class_files.items():
        for ann in getattr(cf, "annotations", []):
            args = ann.get("args", {})
            name_val = args.get("name")
            if isinstance(name_val, str) and name_val.strip():
                by_annotation_type.setdefault(ann.get("type"), []).append((internal, name_val))

    hints = {}
    for ann_type, entries in by_annotation_type.items():
        if len(entries) < 3:
            continue  # мало примеров - может быть случайность, не фреймворк
        for internal, name_val in entries:
            simple = internal.rsplit("/", 1)[-1]
            if not looks_obfuscated_fn(simple, "class"):
                continue  # и так нормальное имя, подсказка не нужна
            sanitized = _sanitize_identifier(name_val)
            if sanitized:
                hints[internal] = sanitized
    return hints


_BRIGADIER_MARKER = b"brigadier"

# aload_0; ldc|ldc_w <idx>; invokespecial <idx2>
_ALOAD_0 = 0x2a
_LDC = 0x12
_LDC_W = 0x13
_INVOKESPECIAL = 0xb7


def _project_uses_brigadier(class_files):
    for cf in class_files.values():
        for entry in cf.pool.values():
            if entry and entry[0] == "Utf8" and _BRIGADIER_MARKER in entry[1].lower().encode("utf-8", "ignore"):
                return True
    return False


def by_brigadier_super_call(class_files, looks_obfuscated_fn):
    if not _project_uses_brigadier(class_files):
        return {}

    hints = {}
    for internal, cf in class_files.items():
        simple = internal.rsplit("/", 1)[-1]
        if not looks_obfuscated_fn(simple, "class"):
            continue
        super_internal = cf.super_class_name
        if not super_internal:
            continue
        super_simple = super_internal.rsplit("/", 1)[-1]
        if not looks_obfuscated_fn(super_simple, "class"):
            continue  # суперкласс не выглядит обфусцированным - не похоже на командный базовый класс
        for m in cf.methods:
            if m.name != "<init>" or not m.code or len(m.code) < 6:
                continue
            code = m.code
            if code[0] != _ALOAD_0:
                continue
            if code[1] == _LDC:
                str_idx = code[2]
                pos = 3
            elif code[1] == _LDC_W:
                str_idx = (code[2] << 8) | code[3]
                pos = 4
            else:
                continue
            if pos + 3 > len(code) or code[pos] != _INVOKESPECIAL:
                continue
            invoke_idx = (code[pos + 1] << 8) | code[pos + 2]
            ref = cf.ref_string(invoke_idx)
            if ref is None or ref[0] != super_internal or ref[1] != "<init>":
                continue
            entry = cf.pool.get(str_idx)
            if not entry or entry[0] != "String":
                continue
            cmd_name = cf.utf8(entry[1])
            if not cmd_name:
                continue
            sanitized = _sanitize_identifier(cmd_name, prefix="Command")
            if sanitized:
                hints[internal] = sanitized
            break  # нашли <init>, дальше методы этого класса не смотрим
    return hints

File: resources/engine_cpp/src/naming_hints.cpp
// naming_hints.cpp - см. naming_hints.hpp. 1:1 порт naming_hints.py.
#include <cstdint>  // БАГ-ФИКС: MinGW/Windows не тянет int64_t транзитивно через другие заголовки, как это молча делает libstdc++ на Linux - см. ошибку сборки Windows-раннера в этой сессии.
#include "naming_hints.hpp"

#include <algorithm>
#include <cctype>

namespace nd {

namespace {

// Превращает произвольную строку в валидный Java-идентификатор в
// PascalCase - 'teleport-player' -> 'TeleportPlayer', 'lobby' -> 'Lobby'.
std::optional<std::string> sanitize_identifier(const std::string& raw, const std::string& prefix = "") {
    std::vector<std::string> parts;
    std::string cur;
    for (char c : raw) {
        bool alnum = std::isalnum(static_cast<unsigned char>(c)) != 0;
        if (alnum) {
            cur += c;
        } else if (!cur.empty()) {
            parts.push_back(cur);
            cur.clear();
        }
    }
    if (!cur.empty()) parts.push_back(cur);
    if (parts.empty()) return std::nullopt;

    std::string name;
    for (auto& p : parts) {
        std::string pp = p;
        pp[0] = static_cast<char>(std::toupper(static_cast<unsigned char>(pp[0])));
        name += pp;
    }
    std::string cleaned;
    for (char c : name)
        if (std::isalnum(static_cast<unsigned char>(c)) || c == '_') cleaned += c;
    name = cleaned;
    if (name.empty()) return std::nullopt;
    if (std::isdigit(static_cast<unsigned char>(name[0]))) name = "_" + name;
    return prefix + name;
}

std::string rsplit_last_slash(const std::string& internal) {
    auto pos = internal.find_last_of('/');
    return pos == std::string::npos ? internal : internal.substr(pos + 1);
}

}  // namespace

std::map<std::string, std::string> hints_by_annotation_name(const std::map<std::string, ClassFile>& class_files,
                                                              const LooksObfuscatedFn& looks_obfuscated_fn) {
    std::map<std::string, std::vector<std::pair<std::string, std::string>>> by_annotation_type;
    for (auto& [internal, cf] : class_files) {
        for (auto& ann : cf.annotations) {
            for (auto& [key, val] : ann.args) {
                if (key != "name" || !val || val->kind != AnnotationValue::Kind::Str) continue;
                std::string s = val->str_v;
                size_t b = s.find_first_not_of(" \t\n\r");
                if (b == std::string::npos) continue;
                by_annotation_type[ann.type].emplace_back(internal, s);
            }
        }
    }

    std::map<std::string, std::string> hints;
    for (auto& [ann_type, entries] : by_annotation_type) {
        (void)ann_type;
        if (entries.size() < 3) continue;
        for (auto& [internal, name_val] : entries) {
            std::string simple = rsplit_last_slash(internal);
            if (!looks_obfuscated_fn(simple, "class")) continue;
            auto sanitized = sanitize_identifier(name_val);
            if (sanitized.has_value()) hints[internal] = *sanitized;
        }
    }
    return hints;
}

namespace {

bool contains_ci(const std::string& haystack_lower, const std::string& needle_lower) { return haystack_lower.find(needle_lower) != std::string::npos; }

bool project_uses_brigadier(const std::map<std::string, ClassFile>& class_files) {
    for (auto& [internal, cf] : class_files) {
        (void)internal;
        for (auto& [idx, entry] : cf.pool) {
            (void)idx;
            if (entry.tag != CpTag::Utf8) continue;
            std::string lower = entry.utf8_value;
            std::transform(lower.begin(), lower.end(), lower.begin(), [](unsigned char c) { return std::tolower(c); });
            if (contains_ci(lower, "brigadier")) return true;
        }
    }
    return false;
}

constexpr uint8_t kAload0 = 0x2a;
constexpr uint8_t kLdc = 0x12;
constexpr uint8_t kLdcW = 0x13;
constexpr uint8_t kInvokespecial = 0xb7;

}  // namespace

std::map<std::string, std::string> hints_by_brigadier_super_call(const std::map<std::string, ClassFile>& class_files,
                                                                   const LooksObfuscatedFn& looks_obfuscated_fn) {
    std::map<std::string, std::string> hints;
    if (!project_uses_brigadier(class_files)) return hints;

    for (auto& [internal, cf] : class_files) {
        std::string simple = rsplit_last_slash(internal);
        if (!looks_obfuscated_fn(simple, "class")) continue;
        if (!cf.super_class_name.has_value()) continue;
        const std::string& super_internal = *cf.super_class_name;
        std::string super_simple = rsplit_last_slash(super_internal);
        if (!looks_obfuscated_fn(super_simple, "class")) continue;

        for (auto& m : cf.methods) {
            if (m.name != "<init>" || !m.has_code || m.code.size() < 6) continue;
            const auto& code = m.code;
            if (code[0] != kAload0) continue;
            uint16_t str_idx;
            size_t pos;
            if (code[1] == kLdc) {
                str_idx = code[2];
                pos = 3;
            } else if (code[1] == kLdcW) {
                str_idx = static_cast<uint16_t>((code[2] << 8) | code[3]);
                pos = 4;
            } else {
                continue;
            }
            if (pos + 3 > code.size() || code[pos] != kInvokespecial) continue;
            uint16_t invoke_idx = static_cast<uint16_t>((code[pos + 1] << 8) | code[pos + 2]);
            auto ref = cf.ref_string(invoke_idx);
            if (!ref.has_value() || std::get<0>(*ref) != super_internal || std::get<1>(*ref) != "<init>") continue;
            auto pit = cf.pool.find(str_idx);
            if (pit == cf.pool.end() || pit->second.tag != CpTag::String) continue;
            auto utf8_it = cf.pool.find(pit->second.idx1);
            if (utf8_it == cf.pool.end() || utf8_it->second.tag != CpTag::Utf8) continue;
            std::string cmd_name = utf8_it->second.utf8_value;
            if (cmd_name.empty()) continue;
            auto sanitized = sanitize_identifier(cmd_name, "Command");
            if (sanitized.has_value()) hints[internal] = *sanitized;
            break;
        }
    }
    return hints;
}

}  // namespace nd
